File: sca-personal-tokens/get-statements-sca.py
import os
import sys
import base64
import json
import rsa
import urllib3
from urllib.parse import urlencode
from datetime import datetime, timedelta, timezone

private_key_path = './private.pem'  # Change to private key path
profile_id = ''                     # Change to profile ID
account_id = ''                     # Change to borderless account ID
base_url = 'https://api.transferwise.com'

if os.getenv('API_TOKEN') is None:
    print('panic: no api token, please set with $ export API_TOKEN=xxx')
    sys.exit(0)
elif profile_id == '' or account_id == '':
    print('panic: profile / account ID missing, please add them')
    sys.exit(0)
elif os.path.exists(private_key_path) is False:
    print('panic: private key file not found, please update key path')
    sys.exit(0)

token = os.getenv('API_TOKEN')  # Set API token from env
http = urllib3.PoolManager()

def get_statement(one_time_token, signature):
    interval_start = (datetime.now(timezone.utc) - timedelta(days=14)).isoformat()
    interval_end = datetime.now(timezone.utc).isoformat()

    params = urlencode({
        'currency': 'GBP', 'type': 'COMPACT',
        'intervalStart': interval_start,
        'intervalEnd': interval_end})

    url = (
        base_url + '/v3/profiles/' + profile_id + '/borderless-accounts/' 
        + account_id + '/statement.json?' + params)

    headers = {
        'Authorization': 'Bearer ' + token,
        'User-Agent': 'tw-statements-sca',
        'Content-Type': 'application/json'}
    if one_time_token != "":
        headers['x-2fa-approval'] = one_time_token
        headers['X-Signature'] = signature
        print(headers['x-2fa-approval'], headers['X-Signature'])

    print('GET', url)

    r = http.request('GET', url, headers=headers, retries=False)

    print('status:', r.status)
    
    if r.status == 200 or r.status == 201:
        return json.loads(r.data)
    elif r.status == 403 and r.getheader('x-2fa-approval') is not None:
        one_time_token = r.getheader('x-2fa-approval')
        signature = do_sca_challenge(one_time_token)
        get_statement(one_time_token, signature)
    else:
        print('failed: ', r.status)
        print(r.data)
        sys.exit(0)

def do_sca_challenge(one_time_token):
    print('doing sca challenge')

    # Read the private key file as bytes.
    with open(private_key_path, 'rb') as f:
        private_key_data = f.read()

    private_key = rsa.PrivateKey.load_pkcs1(private_key_data, 'PEM')

    # Use the private key to sign the one-time-token that was returned 
    # in the x-2fa-approval header of the HTTP 403.
    signed_token = rsa.sign(
        one_time_token.encode('ascii'), 
        private_key, 
        'SHA-256')

    # Encode the signed message as friendly base64 format for HTTP 
    # headers.
    signature = base64.b64encode(signed_token).decode('ascii')

    return signature

def main():
    statement = get_statement("", "")

    if statement is not None and 'currency' in statement['request']:
        currency = statement['request']['currency']
    else:
        print('something is wrong')
        print(statement)
        sys.exit(0)
    
    if 'transactions' in statement:
        txns = len(statement['transactions'])
    else:
        print('Empty statement')
        sys.exit(0)

    print('\n', currency, 'statement received with', txns, 'transactions.')

if __name__ == '__main__':
    main()

File: sca-personal-tokens/get-statements-sca.go
package main

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"time"
)

const (
	privateKeyPath = "./private.pem" // Change to private key path
	profileID      = ""              // Change to profile ID
	accountID      = ""              // Change to borderless account ID
	baseURL        = "https://api.transferwise.com"
)

var token = os.Getenv("API_TOKEN")

func init() {
	if token == "" {
		panic("no api token, please set with $ export API_TOKEN=xxx")
	}
	if profileID == "" || accountID == "" {
		panic("profile / account ID missing, please add them")
	}
	if _, err := os.Stat(privateKeyPath); os.IsNotExist(err) {
		panic("private key file not found, please update key path")
	}
}

func main() {
	currency := "GBP"
	periodDays := 14

	statement, err := getStatement(currency, periodDays)
	if err != nil {
		panic(err)
	}

	fmt.Printf("\n%s statement received with %d transactions.",
		currency, len(statement.Transactions))
}

// Transaction represents a single statement line, that shows a CREDIT
// or DEBIT movement in the account.
type Transaction struct {
	txnType string    `json:'type,omitempty'`
	date    time.Time `json:'date,omitempty'`
}

// Statement is a period of activity in a currency account.
type Statement struct {
	Transactions []Transaction `json:'transactions,omitempty'`
}

func getStatement(currency string, days int) (Statement, error) {
	statement := &Statement{}
	var oneTimeToken, signature string

	intervalStart := time.Now().AddDate(0, 0, -days).Format(time.RFC3339)
	intervalEnd := time.Now().Format(time.RFC3339)

	params := url.Values{}
	params.Add("currency", currency)
	params.Add("type", "COMPACT")
	params.Add("intervalStart", fmt.Sprintf("%s", intervalStart))
	params.Add("intervalEnd", fmt.Sprintf("%s", intervalEnd))

	response, err := doRequest(oneTimeToken, signature, params)
	if err != nil {
		return Statement{}, err
	}
	defer response.Body.Close()
	body, err := ioutil.ReadAll(response.Body)

	err = json.Unmarshal(body, &statement)
	if err != nil {
		return Statement{}, err
	}

	return *statement, err
}

func handleSCA(oneTimeToken string) (string, error) {
	fmt.Println("doing sca challenge")

	privateKey, err := loadPrivateKey(privateKeyPath)
	if err != nil {
		fmt.Errorf("signer is damaged: %v", err)
	}

	// Hashing the x-2fa-approval token with SHA-256.
	hashed := sha256.Sum256([]byte(oneTimeToken))

	// Signing the token with our private key.
	signedToken, err := rsa.SignPKCS1v15(rand.Reader, privateKey,
		crypto.SHA256, hashed[:],
	)
	if err != nil {
		fmt.Errorf("could not sign token: %v", err)
	}

	// Encoding to string to be included in HTTP headers.
	signature := base64.StdEncoding.EncodeToString(signedToken)

	return signature, nil
}

func loadPrivateKey(filePath string) (*rsa.PrivateKey, error) {
	privateKey := &rsa.PrivateKey{}

	// Read private key .pem file into bytes.
	pemBytes, err := ioutil.ReadFile(filePath)
	if err != nil {
		return &rsa.PrivateKey{}, err
	}

	// Decode the key block from the bytes, ignore rest.
	block, _ := pem.Decode(pemBytes)
	if block == nil {
		return &rsa.PrivateKey{}, errors.New("no key found")
	}

	// Detect the key type and parse it into an rsa key object.
	switch block.Type {
	case "RSA PRIVATE KEY":
		privateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
		if err != nil {
			return &rsa.PrivateKey{}, err
		}
		return privateKey, nil
	default:
		return &rsa.PrivateKey{}, fmt.Errorf("unsupported key type %q", block.Type)
	}

	return privateKey, nil
}

func doRequest(oneTimeToken, signature string, params url.Values) (*http.Response, error) {
	u := fmt.Sprintf("%s/v3/profiles/%s/borderless-accounts/%s/statement.json",
		baseURL, profileID, accountID)

	url, err := url.Parse(u)
	if err != nil {
		return nil, err
	}

	url.RawQuery = params.Encode()

	fmt.Printf("\n%s %s\n", http.MethodGet, url.String())

	request, err := http.NewRequest(http.MethodGet, url.String(), nil)
	if err != nil {
		return nil, err
	}

	request.Header.Set("Authorization", "Bearer "+token)
	request.Header.Set("User-Agent", "tw-statements-sca")
	request.Header.Set("Content-Type", "application/json")
	if oneTimeToken != "" {
		request.Header.Set("x-2fa-Approval", oneTimeToken)
		request.Header.Set("X-Signature", signature)
		fmt.Printf("\nx-2fa-approval: %s", request.Header.Get("x-2fa-approval"))
		fmt.Printf("\nX-Signature: %s\n", request.Header.Get("X-Signature"))
	}

	response, err := http.DefaultClient.Do(request)
	if err != nil {
		return nil, err
	}

	fmt.Println(response.StatusCode)

	switch response.StatusCode {
	case http.StatusOK:
		return response, nil
	case http.StatusForbidden:
		if response.Header.Get("x-2fa-approval") != "" {
			oneTimeToken = response.Header.Get("x-2fa-approval")
			signature, err = handleSCA(oneTimeToken)
			if err != nil {
				return nil, err
			}
			return doRequest(oneTimeToken, signature, params)
		}
	default:
		return nil, fmt.Errorf("\nsomething failed, http status: %d", response.StatusCode)
	}

	return response, nil
}
